core/src/main/java/io/grpc/util/RoundRobinLoadBalancerFactory.java

```diff
@@ -211,21 +236,30 @@ public void shutdown() {
      */
     private void updateBalancingState(ConnectivityState state, Status error) {
       List<Subchannel> activeList = filterNonFailingSubchannels(getSubchannels());
-      helper.updateBalancingState(state, new Picker(activeList, error, stickinessState));
+      ready = !activeList.isEmpty();
+      if (state == null) {
+        state = ready ? READY : getAggregatedState();
+      }
+      SubchannelPicker picker = ready ? new Picker(activeList, stickinessState) :
+          new EmptyPicker(error != null ? error : getAggregatedError());
+      helper.updateBalancingState(state, picker);
     }
 
     /**
      * Filters out non-ready subchannels.
      */
     private static List<Subchannel> filterNonFailingSubchannels(
         Collection<Subchannel> subchannels) {
-      List<Subchannel> readySubchannels = new ArrayList<Subchannel>(subchannels.size());
+      List<Subchannel> readySubchannels = null;
       for (Subchannel subchannel : subchannels) {
-        if (getSubchannelStateInfoRef(subchannel).value.getState() == READY) {
+        if (isReady(subchannel)) {
```

Comment:
@zhangkun83 isn't this going to be an expensive linear time operation?  Why is the list scanned each time, rather than Subchannel state change?